src/main/kotlin/com/openlattice/data/storage/DataDeletionService.kt

```diff
@@ -140,38 +136,94 @@ class DataDeletionService(
+                            || edge.src.entitySetId == entitySetId && filteringNeighborEntitySetIds.contains(edge.dst.entitySetId)
+                }
                 .flatMap { edge -> setOf(edge.src, edge.dst) }
                 .groupBy { it.entitySetId }
+                .mapValues { it.value.map { it.entityKeyId }.toSet() }
 
         /*
-         * 2 - delete all entities
+        * 2 - check authorization on all entities
+        */
+
+        /* Check entity set permissions */
+
+        val authorizedPropertiesByEntitySets = getAuthorizedPropertyTypesForDeleteByEntitySet(
+                setOf(entitySetId) + entitySetIdToEntityDataKeys.keys, // include original entity set too
+                deleteType,
+                Optional.empty(),
+                principals
+        )
+
+        /* Check association permissions */
+
+        // only original entity set can be association entity set
+        val isAssociation = entitySetManager.isAssociationEntitySet(entitySetId)
+        val authorizedAssociationPropertiesByEntitySets = if (isAssociation) {
+            // no need to check associations on original entity set
+            getAuthorizedPropertyTypesOfAssociations(entitySetIdToEntityDataKeys, deleteType, principals)
+        } else {
+            getAuthorizedPropertyTypesOfAssociations(
+                    mapOf(entitySetId to entityKeyIds) + entitySetIdToEntityDataKeys,
+                    deleteType,
+                    principals
+            )
+        }
+
+
+        /*
+         * 3 - delete all entities
          */
 
         /* Delete entity */
 
         var numUpdates = 0
 
-        val entityWriteEvent = clearOrDeleteEntities(entitySetId, entityKeyIds, deleteType, principals)
+        val entityWriteEvent = clearOrDeleteAuthorizedEntities(
+                entitySetId,
+                entityKeyIds,
+                deleteType,
+                authorizedPropertiesByEntitySets.getValue(entitySetId)
+        )
         numUpdates += entityWriteEvent.numUpdates
 
-        /* 3 - Delete neighbors */
+        if (isAssociation) {
+            // if entity set is association, we only delete its edges
+            deleteEdgesForAssociationEntitySet(entitySetId, Optional.of(entityKeyIds), deleteType)
+        } else {
+            val associationNumUpdates = clearOrDeleteAuthorizedAssociations(
+                    entitySetId,
+                    Optional.of(entityKeyIds),
+                    deleteType,
+                    authorizedAssociationPropertiesByEntitySets.getValue(entitySetId)
+            )
+            numUpdates += associationNumUpdates
+        }
 
-        // TODO what is/was this for?
-        val neighborDeleteEvents = Lists.newArrayList<AuditableEvent>()
 
-        numUpdates += entitySetIdToEntityDataKeysMap.entries.stream().mapToInt { entry ->
+        /* 3 - Delete neighbors */
```

Comment:
Seems like this will be more performant if we figure out exactly which entities + edges will need to be deleted up front, and then do a bulk delete at the end

Reply:
Maybe I'm missing here something, but I don't think we can do a batch delete here since right now we delete by entity set (and partition) in PEDQS and the `entitySetIdToEntityDataKeysMap` is already mapped by entity set ids.
Or did you mean something else by bulk delete?

Reply:
yeah true, although you could still parallelize by submitting all the clear/delete calls at once instead of calling sequentially and waiting for each one to complete before kicking off the next.

Reply:
👍